File: PA/Laborator/2020-2021/lab01 - Divide et Impera/skel/java/task-3/src/Main.java
import java.io.File;
import java.io.IOException;
import java.io.PrintWriter;
import java.util.Scanner;

public class Main {
    static class Task {
        public final static String INPUT_FILE = "in";
        public final static String OUTPUT_FILE = "out";

        int n, x, y;

        public void solve() {
            readInput();
            writeOutput(getResult(n, x, y));
        }

        private void readInput() {
            try {
                Scanner sc = new Scanner(new File(INPUT_FILE));
                n = sc.nextInt();
                x = sc.nextInt();
                y = sc.nextInt();
                sc.close();
            } catch (IOException e) {
                throw new RuntimeException(e);
            }
        }

        private void writeOutput(int answer) {
            try {
                PrintWriter pw = new PrintWriter(new File(OUTPUT_FILE));
                pw.printf("%d\n", answer);
                pw.close();
            } catch (IOException e) {
                throw new RuntimeException(e);
            }
        }

        private int calcul(int lin, int col, int lat) {
            int r = 0;
            if (lat == 1) {
                return 1;
            }

            lat /= 2;

            // jumatatea de jos
            if (lin > lat) {
                r += 2 * lat * lat;
                lin -= lat;
            }

            // jumatatea din dreapta
            if (col > lat) {
                r += lat * lat;
                col -= lat;
            }
            return r + calcul(lin, col, lat);
        }

        private int getResult(int n, int x, int y) {
            // TODO: Calculati valoarea de pe pozitia (x, y) din matricea de dimensiune 2^n * 2^n.
            return calcul(x, y, (1 << n));
        }
    }

    public static void main(String[] args) {
        new Task().solve();
    }
}

File: PA/Laborator/2020-2021/lab01 - Divide et Impera/skel/cpp/task-3/main.cpp
#include <iostream>
#include <fstream>
using namespace std;

class Task {
public:
    void solve() {
        read_input();
        print_output(get_result(n, x, y));
    }

private:
    int n, x, y;

    void read_input() {
        ifstream fin("in");
        fin >> n >> x >> y;
        fin.close();
    }

    int calcul(int lin, int col, int lat) {
        int r = 0;
        if (lat == 1) {
            return 1;
        }

        lat /= 2;

        // jumatatea de jos
        if (lin > lat) {
            r += 2 * lat * lat;
            lin = lin - lat;
        }

        // jumatatea din dreapta
        if (col > lat) {
            r += lat * lat;
            col = col - lat;
        }

        return r + calcul(lin, col, lat);
    }

    int get_result(int n, int x, int y) {
        /*
        TODO: Calculati valoarea de pe pozitia (x, y) din matricea de dimensiune
        2^n x 2^n
        */
        return calcul(x, y, (1 << n));
    }

    void print_output(int result) {
        ofstream fout("out");
        fout << result;
        fout.close();
    }
};

// Please always keep this simple main function!
int main() {
    // Allocate a Task object on heap in order to be able to
    // declare huge static-allocated data structures inside the class.
    Task *task = new Task();
    task->solve();
    delete task;
    return 0;
}
